### PR TITLE
[DR-2837] Added form components and validation for step 1 of dataset schema creation

### PR DESCRIPTION
**Pull request by lu-c**
Before
Screen Shot 2022-11-16 at 4 04 56 PM

After
Screen Shot 2022-11-23 at 11 36 53 AM

Branch is currently deployed:
https://jade-cl.datarepo-dev.broadinstitute.org/datasets

To use, open the developer tools, and use this `localStorage.setItem('flags', '{"datasetSchemaCreation":{"active":true}}')`. Then refresh the page. The "create dataset" button should now be available on the dataset page

### CODE REVIEW

File: src/components/dataset/overview/DatasetOverviewPanel.jsx
Comment (lu-c) [inline src/components/dataset/overview/DatasetOverviewPanel.jsx]: Not sure if this was a bad delete, but I think this double was the result of a bad merge?

Comment (okotsopoulos) [reply src/components/dataset/overview/DatasetOverviewPanel.jsx]: Nice catch! I had noticed this in dev and meant to file a ticket, thanks for fixing.  A bad merge seems plausible.

Comment (snf2ye) [reply src/components/dataset/overview/DatasetOverviewPanel.jsx]: This was fixed in a [recent PR](https://github.com/DataBiosphere/jade-data-repo-ui/pull/1257/files), right?

File: src/components/dataset/schemaCreation/DatasetSchemaCreationView.tsx
Comment (s-rubenstein) [inline src/components/dataset/schemaCreation/DatasetSchemaCreationView.tsx:126]: Is there a reason to have this function before it does anything? It looks like it is feature flagged, but my general rule tends to be waiting to add the method until implementation time to avoid todo code ending in production.

File: src/components/dataset/schemaCreation/DatasetSchemaCreationView.tsx
Comment (s-rubenstein) [inline src/components/dataset/schemaCreation/DatasetSchemaCreationView.tsx:134]: Same as the comment above, I would say just leave off the submit for now until implementation time.

File: src/components/dataset/schemaCreation/DatasetSchemaCreationView.tsx
Comment (s-rubenstein) [inline src/components/dataset/schemaCreation/DatasetSchemaCreationView.tsx]: I would recommend pulling this out into a component. There's enough display complexity here that it might help with understandability to pull it out. I also think that this is very reusable if you have other tab based forms.

File: src/components/dataset/schemaCreation/DatasetSchemaCreationView.tsx
Comment (s-rubenstein) [inline src/components/dataset/schemaCreation/DatasetSchemaCreationView.tsx:180]: Nit: For the sake of understandability, would you mind changing the order of the props to match above? Just to make it clear the only difference is `type` and `onClick`

File: src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx
Comment (s-rubenstein) [inline src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx:22]: Optional Nit: It looks like generally y'all use absolute paths rather than relative? Should this be an absolute path?

File: src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx
Comment (s-rubenstein) [inline src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx:85]: Just to confirm, is this workable with a screen reader? And how does it behave with keyboard only navigation? (It looks okay, just wanna check to be sure.)

File: src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx
Comment (s-rubenstein) [inline src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx]: Should there be a max length for this?

File: src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx
Comment (s-rubenstein) [inline src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx]: Could "gcp" be abstracted into a constant?

File: src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx
Comment (s-rubenstein) [inline src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx:292]: Nit: What is the `*` afterwards for?

**Comment by s-rubenstein** [comment]
Please ping me when you are ready for a next round!

File: src/components/dataset/schemaCreation/DatasetSchemaCreationView.tsx
Comment (snf2ye) [inline src/components/dataset/schemaCreation/DatasetSchemaCreationView.tsx:206]: Is there a ticket for making this a non-dead link? just don't want to forget!

File: src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx
Comment (snf2ye) [inline src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx:87]: Thoughts on the Capitalization of these headers? I see they match the mocks, but this seems a bit informal. I think we should consider capitalizing the following:
- Dataset Information
- Description (Can we drop "dataset" from this?)
- Terra Project
- Secure Monitoring

File: src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx
Comment (snf2ye) [inline src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx:216]: Since we're selecting from a distinct list of possible regions, I don't think you should be able to edit the region once you've selected one. It's strange to me that once you select one, the cursor is still present and you can backspace letters from the region. This is different than how the Cloud Platform option is working. 

https://user-images.githubusercontent.com/13254229/205720985-12e85ed2-1f5b-402d-9716-39cfe046073a.mov

File: src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx
Comment (snf2ye) [inline src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx:100]: Can we also add placeholders for the other fields?
- description
- Terra Project
- Region
- Steward/custodian list

File: src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx
Comment (snf2ye) [inline src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx]: Edge case: If the user changes the region, we should also clear any value already set in the region field.

File: src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx
Comment (snf2ye) [inline src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx:280]: I don't understand this pattern - why are we calling onChange from within the onChange function?

File: src/constants/index.ts
Comment (snf2ye) [inline src/constants/index.ts:217]: What do the "count" and "letters" options do?

File: src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx
Comment (snf2ye) [inline src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx]: nit: Capitalization Consistency - Can we either capitalize "name" for the required error message or make the rest of "Name" options lower case?

File: src/components/dataset/schemaCreation/DatasetSchemaCreationView.tsx
Comment (lu-c) [reply src/components/dataset/schemaCreation/DatasetSchemaCreationView.tsx:126]: No *good* reason 😓 I had it in there initially so I could have the form behavior (with the react-hook-form library) set up, just to confirm the form behavior was working as intended, and was passing the right values to the submission obj.

I hope its okay if we overlook this for now just because of the feature flag, and the implementation has been added already in the follow up branch!

File: src/components/dataset/schemaCreation/DatasetSchemaCreationView.tsx
Comment (lu-c) [reply src/components/dataset/schemaCreation/DatasetSchemaCreationView.tsx:134]: (see above)

File: src/components/dataset/schemaCreation/DatasetSchemaCreationView.tsx
Comment (lu-c) [reply src/components/dataset/schemaCreation/DatasetSchemaCreationView.tsx]: Would you mind clarifying which part should be a new component? I have it set up rn so the tab contents are separate components... do you mean the tabs should be its own one too?

File: src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx
Comment (lu-c) [reply src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx]: Oh shoot, I didnt notice there was a maxlen for this in the mock up!!! Ok, fixing!

File: src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx
Comment (lu-c) [reply src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx]: Updated! There was already a constant for this! :D

File: src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx
Comment (lu-c) [reply src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx:292]: the * means the field is required to have a value

File: src/components/dataset/schemaCreation/DatasetSchemaCreationView.tsx
Comment (lu-c) [reply src/components/dataset/schemaCreation/DatasetSchemaCreationView.tsx:206]: No, let me create one now 👍

Comment (lu-c) [reply src/components/dataset/schemaCreation/DatasetSchemaCreationView.tsx:206]: https://broadworkbench.atlassian.net/browse/DR-2859

File: src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx
Comment (lu-c) [reply src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx:87]: I prefer to capitalize all first letters in any titles or labels too, but I think this is the route that other apps have done? Like terra. So maybe it's better for consistency?

File: src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx
Comment (lu-c) [reply src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx]: Yes! Let me capitalize those!

File: src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx
Comment (lu-c) [reply src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx:216]: ok! I wanted this to be different actually, because the cloud platform only has 2 options, so a select is easy. But when the possible regions are > 100, i didn't want the user to have to scroll through the list, and wanted to give them the ability to type the answer in.

I removed the flag `freeSolo` that allowed that component to allow any value (so now it is constrained to the known options list), but still allows the user to type.

Does this help? Let me know if i should readjust!

File: src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx
Comment (lu-c) [reply src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx:100]: Added placeholders for everything but the terra project field - Going to hold off on that one until the follow up PR, if that's okay! The 2nd PR will fix the terra project field so it auto populates the first one, but it has to make a backend call to get them.

File: src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx
Comment (lu-c) [reply src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx]: Fixed! :)

File: src/constants/index.ts
Comment (lu-c) [reply src/constants/index.ts:217]: I just got these values from the GCP region list in google's documetnation for what the server options are. There are a lot of "asia-east1-a" "asia-east2-a", etc. Instead of copying them all in to a giant variable, I compacted it a little so we generate them instead.
https://cloud.google.com/compute/docs/regions-zones

Unfortunately, this list also changes in the follow up PR too! I didn't realize jade-repo is not expecting letters at all...

File: src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx
Comment (lu-c) [reply src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx:22]: (@snf2ye What do you think?)

I haven't paid much attention to the imports, but I assumed this was okay?

File: src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx
Comment (lu-c) [reply src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx:280]: There are nested components here, so react-hook-forms is wrapping form elements. It knows how to handle updating the form value for default components, but then struggles with more complex methods. They let you have access to the underlying methods when you wrap it in the Controller object, but then you have 2 onChange methods.

The first onChange is from mui/autocomplete. The second onchange being called belongs to the react-hook-form/controller

File: src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx
Comment (lu-c) [reply src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx:85]: Tabbing through, it looks okay to me!

There is one area that sucks for keyboard navigation though, the description markdown editor treats the tabs as characters. Not sure how to resolve that right now, but I will add this as a bug ticket!

Otherwise, it seems okay!

File: src/components/dataset/overview/DatasetOverviewPanel.jsx
Comment (lu-c) [reply src/components/dataset/overview/DatasetOverviewPanel.jsx]: Ah yes! Pulled from develop ✨

File: src/components/dataset/schemaCreation/DatasetSchemaCreationView.tsx
Comment (s-rubenstein) [reply src/components/dataset/schemaCreation/DatasetSchemaCreationView.tsx:126]: Sure, if that is team agreement, that's probably okay.

File: src/components/dataset/schemaCreation/DatasetSchemaCreationView.tsx
Comment (s-rubenstein) [reply src/components/dataset/schemaCreation/DatasetSchemaCreationView.tsx]: I think that the ternary content seems fairly standard (with the stepping through tabs). Because of that, I would pull the div out that starts on line 161.

File: src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx
Comment (s-rubenstein) [reply src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx:87]: I think the Terra UX has first letter only. I also prefer first letter of each word. Might be worth checking with UX to see if it is intentional?

File: src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx
Comment (s-rubenstein) [reply src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx:292]: TOL: I wonder if that would be doable with a subcomponent that takes a prop named `required` and automatically adds that `*`. It feels a little strange to put that in the text, because I think it could easily lead to inconsistencies in the way required fields are handled across the UI.

File: src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx
Comment (s-rubenstein) [inline src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx]: Do you think that these validation rules could be reusable in other forms? Did you think about pulling them out into a helper file so that other forms could make use of them? (Especially ones like this one that have some meat, but are going to be identical for many email validation checks)

File: src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx
Comment (s-rubenstein) [inline src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx]: (See above comment, this is actually an example where that reusability would be very powerful.)

File: src/libs/render-utils.tsx
Comment (s-rubenstein) [inline src/libs/render-utils.tsx]: For a potential improvement for the future, I wonder if you could put the regions in this constant? Then you could do some filtering and mapping to get regions for a particular cloudPlatform, rather than needing to do conditionals. I think that might be more powerful if you end up needing to add another cloud platform to simplify the number of places you would need to add. (I think you might be able to get it to the point where the only place you would need to modify is here, and every bit of the UI would automatically update to use that new cloudPlatform)

File: src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx
Comment (lu-c) [reply src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx:87]: Lou confirms they want to only capitalize the first letter of the first word

File: src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx
Comment (lu-c) [reply src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx]: Ahm, yeah, i can pull it out into a helper file

File: src/libs/render-utils.tsx
Comment (lu-c) [reply src/libs/render-utils.tsx]: The future is now! updated this by moving the constant out of the utils and into the CONSTANTS file where it belongs!

File: src/components/dataset/schemaCreation/DatasetSchemaCreationView.tsx
Comment (lu-c) [reply src/components/dataset/schemaCreation/DatasetSchemaCreationView.tsx]: Turns out it already exists as a component! 🎉

File: src/components/dataset/schemaCreation/DatasetSchemaCreationView.tsx
Comment (s-rubenstein) [inline src/components/dataset/schemaCreation/DatasetSchemaCreationView.tsx]: This could also use the constant.

File: src/constants/index.ts
Comment (s-rubenstein) [reply src/constants/index.ts:217]: Could you add that as a comment? I think that is valuable information (showing what the compound version is)

File: src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx
Comment (s-rubenstein) [inline src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx:145]: It feels like a lot of the Grid item code is repeated. I wonder if that could be an abstracted component? Especially because this is a very meaty component right now, and I think it could easily end up with high complexity as things need to change.

**Review by s-rubenstein — COMMENTED** [review]
I think trying to break up the `DatasetSchemaInformationView` component is the last thing that I would try to change! After that it should be good to go.

File: src/components/dataset/schemaCreation/DatasetSchemaCreationView.tsx
Comment (lu-c) [reply src/components/dataset/schemaCreation/DatasetSchemaCreationView.tsx]: updated!

File: src/constants/index.ts
Comment (lu-c) [reply src/constants/index.ts:217]: done!

File: src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx
Comment (lu-c) [reply src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx:145]: I am going to push back on that a little? Grid is already a component, I don't really see too much value in abstracting it out further, we would be trading lines of code for lines of config imo. I prefer not to make components unless they are really reuseable as well, and I don't think a Grid surrounding a single component will be

Comment (s-rubenstein) [reply src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx:145]: My biggest concern is the repeated code of the children. For example:
```
<label
          htmlFor="{FIELD}"
          className={clsx(classes.formLabel, { [classes.formLabelError]: {ERROR CLASS} })}
        >
          {LABEL DISPLAY}
        </label>
```
Is almost verbatim repeated across every grid component. However, especially since its not my codebase, if you think that the value gain is negligible, I am totally okay being overruled here. I just always find it a bit of a yellow flag when I see a lot of repeated code like this. (Another option would be to pull form fields out into a list of objects, and then map over that list to create the grid elements, but I think I also defer to the folks working in the codebase for style preferences)

**Review by s-rubenstein — APPROVED** [review]

File: src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx
Comment (lu-c) [reply src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx:145]: In this case, I also want to defer to the tech leads in the project too 😂 I'm also interested in hearing other opinions too! Maybe I'm off base here! :)

File: src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx
Comment (snf2ye) [reply src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx:216]: Gotcha! Yes, I like that you can't actually submit an invalid region now.

**Review by snf2ye — APPROVED** [review]

**Pull request merged**
